pkg/driver/bigquery/bigquery.go

```diff
@@ -0,0 +1,525 @@
+			return false, err
+		}
+		return true, nil
+	}
+
+	exists, err := drv.DatabaseExists()
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		_, err = datasetDrop(*drv.context, drv.client)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (drv *Driver) DumpSchema(db *sql.DB) ([]byte, error) {
+	var script []byte
+
+	query := fmt.Sprintf(`
+		SELECT
+			table_name,
+			table_type,
+			1 AS order_type
+		FROM
+			`+"`%s.%s.INFORMATION_SCHEMA.TABLES`"+`
+		WHERE
+			table_type = 'BASE TABLE'
+		UNION ALL
+		SELECT
+			table_name,
+			table_type,
+			2 AS order_type
+		FROM
+			`+"`%s.%s.INFORMATION_SCHEMA.TABLES`"+`
+		WHERE
+			table_type = 'VIEW'
+		UNION ALL
+		SELECT
+			routine_name AS table_name,
+			'FUNCTION' AS table_type,
+			3 AS order_type
+		FROM
+			`+"`%s.%s.INFORMATION_SCHEMA.ROUTINES`"+`
+		ORDER BY
+			order_type;`, drv.projectID, drv.datasetID, drv.projectID, drv.datasetID, drv.projectID, drv.datasetID)
+
+	// Execute the query
+	rows, err := db.Query(query)
+	if err != nil {
+		return nil, fmt.Errorf("error querying objects: %v", err)
+	}
+	defer dbutil.MustClose(rows)
+
+	// Iterate over the results and generate DDL for each object
+	for rows.Next() {
+		var objectName, objectType string
+		var orderType int
+		if err := rows.Scan(&objectName, &objectType, &orderType); err != nil {
+			return nil, fmt.Errorf("error scanning object: %v", err)
+		}
+
+		// Generate DDL for the object
+		ddl, err := generateDDL(db, drv.projectID, drv.datasetID, objectName, objectType)
+		if err != nil {
+			return nil, fmt.Errorf("error generating DDL for %s %s: %v", objectName, objectType, err)
+		}
+
+		// Append the DDL to the script
+		script = append(script, []byte(ddl)...)
+		script = append(script, []byte("\n\n")...)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating objects: %v", err)
+	}
+
+	return script, nil
+}
+
+func (drv *Driver) MigrationsTableExists(*sql.DB) (bool, error) {
+	exists, err := tableExists(*drv.context, drv.client, drv.datasetID, drv.migrationsTableName)
+	if err != nil {
+		return exists, err
+	}
+
+	return exists, nil
+}
+
+func (drv *Driver) DeleteMigration(db dbutil.Transaction, version string) error {
+	query := fmt.Sprintf("DELETE FROM %s.%s WHERE version = '%s';", drv.datasetID, drv.migrationsTableName, version)
+	_, err := db.Exec(query)
+
+	return err
+}
+
+func (drv *Driver) InsertMigration(db dbutil.Transaction, version string) error {
+	queryTemplate := `INSERT INTO %s.%s (version) VALUES ('%s');`
+	queryString := fmt.Sprintf(queryTemplate, drv.datasetID, drv.migrationsTableName, version)
+
+	_, err := db.Exec(queryString, version)
+	return err
+}
+
+func (drv *Driver) Open() (*sql.DB, error) {
+	con, err := sql.Open("bigquery", drv.databaseURL)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+	client, err := getClient(ctx, drv.config)
+	if err != nil {
+		return nil, err
+	}
+
+	drv.client = client
+	drv.context = &ctx
+
```

Comment:
Now that you've incorporated the previous feedback, this is actually unnecessary.

`database/sql` provides a mechanism to get the underlying connection, through [func (*DB) Conn](https://pkg.go.dev/database/sql#DB.Conn).

If my understanding is correct, the underlying `go-gorm/bigquery` driver will give you back a `DB.Conn` that is this struct:

https://github.com/go-gorm/bigquery/blob/d4291d2dbfb0501df9f2221f625a34f6f881ef29/driver/connection.go#L10-L17

Inside it is the `*bigquery.Client` connection, etc.  We should use that and not our own private connection, so that all the connection management is handled by `database/sql`, including connection pooling, etc.

Reply:
Unfortunately, I was not able to find any way to get the underlying `bigQueryConnection` which has the `client` field. I tried a bit with reflection but was unsuccessful. `bigQueryConnection` is not exported from the gorm package as well. I am not an expert in Go but if you have any suggestions, I am all ears. Thanks.

Reply:
Using `*sql.Conn`'s [Raw()](https://pkg.go.dev/database/sql#Conn.Raw) gets us a handle to the underlying `bigQueryConnection`, but unfortunately the `go-gorm/bigquery` driver didn't make the `*bigquery.Client` struct member public. 😢

On top of that oversight, for some reason, they implemented `GetDataset()` and a `GetContext()` but no `GetClient()` -- they stopped just short of success.

I know there's all kinds of warnings around digging into structures and accessing unexported fields with reflection, but this might be the way to go, here.  Look at this test:

```
func TestGetClient(t *testing.T) {
	drv := testBigQueryDriver(t)

	db, err := drv.Open()
	require.NoError(t, err)
	defer dbutil.MustClose(db)

	conn, err := db.Conn(context.Background())
	require.NoError(t, err)
	defer conn.Close()

	err = conn.Raw(func(driverConn any) error {
		value := reflect.ValueOf(driverConn).Elem().FieldByName("client")
		value = reflect.NewAt(value.Type(), unsafe.Pointer(value.UnsafeAddr())).Elem()

		client := value.Interface().(*bigquery.Client)
		require.Equal(t, "test", client.Project())

		return nil
	})
	require.NoError(t, err)
}
```

Does that help?

Reply:
works like a charm!